go/src/lib.rs

```diff
@@ -1,9 +1,160 @@
+///
+/// The returned `ConnectionResponse` should be manually freed by calling `free_connection_response`, otherwise a memory leak will occur. It should be freed whether or not an error occurs.
+///
+/// # Safety
+///
+/// * `connection_request_bytes` must point to `connection_request_len` consecutive properly initialized bytes.
+/// * `connection_request_len` must not be greater than `isize::MAX`. See the safety documentation of [`std::slice::from_raw_parts`](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html).
+#[no_mangle]
+pub unsafe extern "C" fn create_client(
+    connection_request_bytes: *const u8,
+    connection_request_len: usize,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> *const ConnectionResponse {
+    let request_bytes =
+        unsafe { std::slice::from_raw_parts(connection_request_bytes, connection_request_len) };
+    let response = match create_client_internal(request_bytes, success_callback, failure_callback) {
+        Err(err) => ConnectionResponse {
+            conn_ptr: std::ptr::null(),
+            error_message: CString::into_raw(CString::new(err.message).unwrap()),
+            error_type: err.error_type,
+        },
+        Ok(client) => ConnectionResponse {
+            conn_ptr: Box::into_raw(Box::new(client)) as *const c_void,
+            error_message: std::ptr::null(),
+            error_type: RequestErrorType::Unspecified,
+        },
+    };
+    Box::into_raw(Box::new(response))
+}
+
+/// Closes the given client, deallocating it from the heap.
+///
+/// # Safety
+///
+/// * `client_ptr` must be able to be safely casted to a valid `Box<Client>` via `Box::from_raw`. See the safety documentation of [`std::boxed::Box::from_raw`](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.from_raw).
+/// * `client_ptr` must not be null.
+#[no_mangle]
+pub unsafe extern "C" fn close_client(client_ptr: *const c_void) {
+    let client_ptr = unsafe { Box::from_raw(client_ptr as *mut Client) };
+    let _runtime_handle = client_ptr.runtime.enter();
+    drop(client_ptr);
+}
+
+/// Deallocates a `ConnectionResponse`.
+///
+/// # Safety
+///
+/// * `connection_response_ptr` must be able to be safely casted to a valid `Box<ConnectionResponse>` via `Box::from_raw`. See the safety documentation of [`std::boxed::Box::from_raw`](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.from_raw).
+/// * `connection_response_ptr` must not be null.
+#[no_mangle]
+pub unsafe extern "C" fn free_connection_response(
+    connection_response_ptr: *const ConnectionResponse,
+) {
+    let connection_response =
+        unsafe { Box::from_raw(connection_response_ptr as *mut ConnectionResponse) };
+    drop(connection_response);
```

Comment:
Don't you need to release `connection_response.error_message` here? I think you can call `free_error`

Reply:
We need to be able to free the connection response separately from the error contained within it. When we get back a connection pointer, the error is null, so when we try to free the null pointer, we get a seg fault.

Reply:
Can you add ?
```rs
if connection_response.error_message != std::ptr::null()
    free_error(connection_response.error_message);
```

Reply:
Okay. I'll do this and also make free_error private instead of exposing it to the wrapper language.